src/iptvsimple/CatchupController.cpp

```diff
@@ -0,0 +1,436 @@
+
+  Logger::Log(LEVEL_DEBUG, "default_url - %s", channel.GetStreamURL().c_str());
+  Logger::Log(LEVEL_DEBUG, "playback_as_live - %s", playbackAsLive ? "true" : "false");
+  Logger::Log(LEVEL_DEBUG, "catchup_url_format_string - %s", GetCatchupUrlFormatString(channel).c_str());
+  Logger::Log(LEVEL_DEBUG, "catchup_buffer_start_time - %s", std::to_string(m_catchupStartTime).c_str());
+  Logger::Log(LEVEL_DEBUG, "catchup_buffer_end_time - %s", std::to_string(m_catchupEndTime).c_str());
+  Logger::Log(LEVEL_DEBUG, "catchup_buffer_offset - %s", std::to_string(m_timeshiftBufferOffset).c_str());
+  Logger::Log(LEVEL_DEBUG, "timezone_shift - %s", std::to_string(channel.GetTvgShift()).c_str());
+}
+
+void CatchupController::TestAndStoreStreamType(Channel& channel)
+{
+  // We need to find out what type of stream this is, so let's construct a catchup URL to test with
+  std::string streamTestUrl = GetStreamTestUrl(channel);
+  StreamType streamType = StreamUtils::GetStreamType(streamTestUrl, channel);
+  if (streamType == StreamType::OTHER_TYPE)
+    streamType = StreamUtils::InspectStreamType(streamTestUrl);
+
+  // TODO: we really want to store this in a file and load it on any restart
+  // using channel doesn't make sense as it's otherwise immutable.
+
+  // If we find a mimetype store it so we don't have to look it up again
+  if (!channel.GetProperty("mimetype").empty() && (streamType == StreamType::HLS || streamType == StreamType::DASH))
+  {
+    std::lock_guard<std::mutex> lock(*m_mutex);
+    channel.AddProperty("mimetype", StreamUtils::GetMimeType(streamType));
+  }
+
+  if (StreamUtils::GetEffectiveInputStreamClass(streamType, channel) == "inputstream.ffmpegdirect")
+    m_controlsLiveStream = true;
+  else
+    m_controlsLiveStream = false;
+}
+
+void CatchupController::UpdateProgrammeFrom(const EPG_TAG& epgTag, int tvgShift)
+{
+  m_programmeStartTime = epgTag.startTime;
+  m_programmeEndTime = epgTag.endTime;
+  m_programmeTitle = epgTag.strTitle;
+  m_programmeUniqueChannelId = epgTag.iUniqueChannelId;
+  m_programmeChannelTvgShift = tvgShift;
+}
+
+void CatchupController::UpdateProgrammeFrom(const data::EpgEntry& epgEntry, int tvgShift)
+{
+  m_programmeStartTime = epgEntry.GetStartTime();
+  m_programmeEndTime = epgEntry.GetEndTime();
+  m_programmeTitle = epgEntry.GetTitle();
+  m_programmeUniqueChannelId = epgEntry.GetChannelId();
+  m_programmeChannelTvgShift = tvgShift;
+}
+
+void CatchupController::ClearProgramme()
+{
+  m_programmeStartTime = 0;
+  m_programmeEndTime = 0;
+  m_programmeTitle.clear();
+  m_programmeUniqueChannelId = 0;
+  m_programmeChannelTvgShift = 0;
+}
+
+// TODO: Need the equivalent of this somehow
+// void CatchupController::CloseLiveStream()
+// {
+//   m_resetCatchupState = true;
+//   Logger::Log(LEVEL_NOTICE, "CloseLiveStream - true");
+// }
+
+namespace
+{
+
+void FormatOffset(time_t tTime, std::string &urlFormatString)
+{
+  const std::string regexStr = ".*(\\{offset:(\\d+)\\}).*";
```

Comment:
"const std::string directly in this unnamed namespace" to not construct it on every call?

Reply:
Ya, so really this should be static regex as regex's are expensive to create.

So if `static const std::regex rx(".*(\\{offset:(\\d+)\\}).*");` the same as `const std::regex rx(".*(\\{offset:(\\d+)\\}).*");`?

Reply:
I went with const without the static.

Reply:
If you want to prevent that the regex is created each time the function is called -- and this should be the goal here -- you either need to make it static inside the function or to put it outside the function in an unnamed namespace.

Reply:
Ah so like?:

```
const std::regex rx(".*(\\{offset:(\\d+)\\}).*");

void FormatOffset(time_t tTime, std::string &urlFormatString)
{
  std::cmatch mr;
  if (std::regex_match(urlFormatString.c_str(), mr, rx) && mr.length() >= 3)
  {
    std::string offsetExp = mr[1].first;
    std::string second = mr[1].second;
```

Reply:
Done!

Reply:
👍

Reply:
I would give the constant a more speaking (because it is visible inside the whole unnamend namespace now) and all upper-case name (because this is common practice for constant's name).

Reply:
On second thought, static inside the function is imo better, because you don't need the value outside that function. unnamed namespace const would be the choice if you need the value even from outside the function.

Reply:
Ok, done, should be good now.